tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -122,9 +118,90 @@ def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedE
         """Event triggered when the read/write endpoints of the database change."""
         logger.info(f"second database endpoints have been changed to: {event.endpoints}")
 
-    def _on_run_query_action(self, event: ActionEvent):
-        """Runs queries."""
-        raise NotImplementedError
+    # ==============
+    #  Action hooks
+    # ==============
+
+    def _on_single_put_action(self, event: ActionEvent):
+        logger.info(event.params)
+        relation = self.first_database
+        relation_id = event.params["relation-id"]
+        databag = relation.fetch_relation_data()[relation_id]
+        method = "PUT"
+        payload = {"artist": "Vulfpeck", "genre": ["Funk", "Jazz"], "title": "Thrill of the Arts"}
```

Comment:
**suggestion** maybe it would be best to put the payload as parameter of the action. This way one could:

1. parametrize the call and use this action in multiple context
2. have this in the test itself, which makes it easier to understand the test in the first place: in the test we can supply certain json, run the action, and do assertion based on the input json. Now this is less clear as the `single_put_action` does some data loading which (when I read the tests) I somewhat obscure and I need to go to the action implementation to understand the data I have inserted (and based on which assertions are done)

Reply:
This was actually the original plan, but I ran into a bug with how JSON is parsed in actions somewhere, and I just couldn't get it through the action without it being mangled, especially when sending a multi-line bulk JSON request. I'll give it another go, since I feel your suggestion would be simpler, but I don't see a problem with having a 1-1 mapping of actions to tests in test charms.

Reply:
I've updated all actions to be a new action called run_request, which runs HTTP requests with a given JSON payload at a set endpoint. Turns out `re.escape()` comes in handy for escaping characters properly